src/compose/app.ts

```diff
@@ -444,45 +539,50 @@ export class App {
 	// it should be changed.
 	private generateStepsForComponent<T extends Volume | Network>(
 		components: Array<ChangingPair<T>>,
-		changingServices: Array<ChangingPair<Service>>,
-		dependencyFn: (component: T, service: Service) => boolean,
+		dependentServices: Service[],
 	): CompositionStep[] {
 		if (components.length === 0) {
 			return [];
 		}
 
 		let steps: CompositionStep[] = [];
 
+		const componentIsVolume =
+			(components[0].current ?? components[0].target) instanceof Volume;
+
 		const actions: {
 			create: CompositionStepAction;
 			remove: CompositionStepAction;
-		} =
-			(components[0].current ?? components[0].target) instanceof Volume
-				? { create: 'createVolume', remove: 'removeVolume' }
-				: { create: 'createNetwork', remove: 'removeNetwork' };
+		} = componentIsVolume
+			? { create: 'createVolume', remove: 'removeVolume' }
+			: { create: 'createNetwork', remove: 'removeNetwork' };
 
 		for (const { current, target } of components) {
 			// If a current exists, we're either removing it or updating the configuration. In
-			// both cases, we must remove the component first, so we output those steps first.
+			// both cases, we must remove the component before creating it to avoid
+			// Engine conflicts. So we always emit a remove step first.
 			// If we do remove the component, we first need to remove any services which depend
-			// on the component
+			// on the component. The service removal steps are generated in this.generateStepsForService
+			// after their removal is calculated in this.compareServices.
 			if (current != null) {
-				// Find any services which are currently running which need to be killed when we
-				// recreate this component
-				const dependencies = _.filter(this.services, (s) =>
-					dependencyFn(current, s),
-				);
-				if (dependencies.length > 0) {
-					// We emit kill steps for these services, and wait to destroy the component in
-					// the next state application loop
-					// FIXME: We should add to the changingServices array, as we could emit several
-					// kill steps for a service
-					steps = steps.concat(
-						dependencies.flatMap((svc) =>
-							this.generateKillStep(svc, changingServices),
-						),
-					);
-				} else {
+				// If there are any dependent services which have the volume or network,
+				// we cannot proceed to component removal.
+				const dependentServicesOfComponent = dependentServices.filter((s) => {
+					if (componentIsVolume) {
+						return this.serviceHasNetworkOrVolume(
+							s,
+							[],
+							components as Array<ChangingPair<Volume>>,
```

Comment:
Why pass the full component array here? Should this be 
```suggestion
							[component],
```

Reply:
Done [here](https://github.com/balena-os/balena-supervisor/pull/2234/commits/10eaf8f347cc5a3536db45e9ec04fa9981be4f23#diff-857bb737cdeb3c33e2caa352769d3f96e0e8a48da6fea254eb99bc627732cec1R575), also agreed, it's not a great function to work with